### PR TITLE
Add GitHub Options

### PR DESCRIPTION
**Pull request by RehanSaeed**
- Adds the following new options:
  - **Contact** - The contact details to use if someone wants to contact you about a security vulnerability or code of conduct issues.
  - **SourceControl** - Select which source control provider you are using if any, to add provider specific features.
    - **GitHub** (Default) - Adds .github directory containing a code of conduct, contributing guide, pull request template and issue templates.
    - **None** - No source control provider is being used.
  - **GitHubActions** (Default=On) - Adds GitHub Actions continuation integration and automatic release drafting.

### CODE REVIEW

**Comment by RehanSaeed** [comment]
I used the API template to create https://github.com/RehanSaeed/FastestApi to test the GitHub Action. It all works and produces a Docker image targetting x64 and ARM64 but fails to push the image to the Docker Hub. Looks like I'm hitting issue https://github.com/moby/buildkit/issues/838.

**Comment by VictorioBerra** [comment]
This looks great. Am I still missing where the .gitignore is though?

**Comment by RehanSaeed** [comment]
It's in there e.g.

https://github.com/Dotnet-Boxed/Templates/blob/master/Source/ApiTemplate/.gitignore

File: Docs/API.md [hunk 1]
```diff
@@ -26,12 +26,23 @@ The ASP.NET Core API project template uses `dotnet new` to enable you to turn fe
...
 
+#### Source Control
+
+- **SourceControl** - Select which source control provider you are using if any, to add provider specific features.
+  - **GitHub** (Default) - Adds .github directory containing a code of conduct, contributing guide, pull request template and issue templates.
+  - **None** - No source control provider is being used.
+
+#### Continuous Integration (CI)
+
+- **GitHubActions** (Default=On) - Adds GitHub Actions continuation integration and automatic release drafting.
```
Comment (VictorioBerra) [inline Docs/API.md:44]: continuation -> continuous

File: Source/ApiTemplate/.github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    - name: 'Dotnet Cake Publish'
+      run: dotnet cake --target=Publish
+      shell: pwsh
+    - name: 'Publish Artefacts'
+      uses: actions/upload-artifact@v1.0.0
+      with:
+        name: ${{matrix.os}}
+        path: './Artefacts'
+
+  docker:
```
Comment (RehanSaeed) [inline Source/ApiTemplate/.github/workflows/build.yml:59]: Wrap this in `#if Docker`.

File: Docs/API.md (hunk 1 above)
Comment (RehanSaeed) [reply Docs/API.md:44]: Good catch. Fixed.

**Comment by RehanSaeed** [comment]
I got it working!!!

https://hub.docker.com/repository/docker/muhammadrehansaeed/api

I'm finally going to start on the Hot Chocolate migration now. Will blog about this soon. It'll be a similar blog post to the NuGet package one I did recently.

**Comment by VictorioBerra** [comment]
Nice job! Checking it out now.

**Comment by VictorioBerra** [comment]
> I got it working!!!
> 
> https://hub.docker.com/repository/docker/muhammadrehansaeed/api

What was the issue? Was it the order of the arguments? The format of the tag maybe?

**Comment by RehanSaeed** [comment]
It was me being dumb. I forgot that when pushing to DockerHub, you need to create a Docker image tag with the format `username/name:version`. To facilitate this, I had to add a new `--tag` argument for the cake script and pass it from the GitHub Action yml.

I found one issue with the dotnet templating. The Docker tag name has to be lower case and you can't transform the `--name` parameter to `dotnet new` to lowercase for that. See https://github.com/dotnet/templating/issues/2210.

One possible improvement I can make is to add a Docker image build cache to GitHub Actions as described in https://github.com/marketplace/actions/docker-buildx#leverage-buildx-cache, so the build will run quicker. Not sure it's worth it though as it does complicate things.

**Comment by VictorioBerra** [comment]
I am struggling a bit. I had to google this error:

> auto-push is currently not implemented for docker driver

And then I ran this `docker buildx create --use --name build --node build --driver-opt network=host` (which im not sure what that even does)

It worked though, and the docker stuff started working. Then I ran:

- `docker login`
- `dotnet tool restore`
- `dotnet cake --push=true`

 and finally I get this:

```
...

 => [linux/arm64 sdk  7/10] COPY . .                                                                                                                                                                                 0.1s 
 => [linux/amd64 sdk  8/10] RUN dotnet build --configuration Release --no-restore                                                                                                                                    9.0s 
 => [linux/arm64 sdk  8/10] RUN dotnet build --configuration Release --no-restore                                                                                                                                    9.1s
 => [linux/amd64 sdk  9/10] RUN dotnet test --configuration Release --no-build                                                                                                                                       7.9s 
 => [linux/arm64 sdk  9/10] RUN dotnet test --configuration Release --no-build                                                                                                                                       7.7s 
 => [linux/arm64 sdk 10/10] RUN dotnet publish "Source/BoxedTemplatesApiTest1/BoxedTemplatesApiTest1.csproj" --configuration Release --no-build --output /app                                                        2.1s 
 => [linux/amd64 sdk 10/10] RUN dotnet publish "Source/BoxedTemplatesApiTest1/BoxedTemplatesApiTest1.csproj" --configuration Release --no-build --output /app                                                        2.0s 
 => CACHED [linux/arm64 runtime 2/4] RUN apk add --no-cache     icu-libs     tzdata                                                                                                                                  0.0s
 => CACHED [linux/arm64 runtime 3/4] WORKDIR /app                                                                                                                                                                    0.0s 
 => CACHED [linux/arm64 runtime 4/4] COPY --from=sdk /app .                                                                                                                                                          0.0s 
 => CACHED [linux/amd64 runtime 2/4] RUN apk add --no-cache     icu-libs     tzdata                                                                                                                                  0.0s 
 => CACHED [linux/amd64 runtime 3/4] WORKDIR /app                                                                                                                                                                    0.0s 
 => CACHED [linux/amd64 runtime 4/4] COPY --from=sdk /app .                                                                                                                                                          0.0s 
 => ERROR exporting to image                                                                                                                                                                                         2.9s 
 => => exporting layers                                                                                                                                                                                              0.1s 
 => => exporting manifest sha256:03a6c086ce68ba0e51cb10bf376a8bd24cf6d08abd640d3b93fa841e2696862d                                                                                                                    0.0s 
 => => exporting config sha256:e0a9b32c21d1aa79e8254b228da0162905f786a9f41fa3a7f7d02f22dc2719ca                                                                                                                      0.0s 
 => => exporting manifest sha256:e99426b13d6b1ffef6b9164e877db53d1677b13063bb329ae8cbc655d20f4ebc                                                                                                                    0.0s 
 => => exporting config sha256:472848340426f5196e31b078b6cee62aeb44e724623cdf3130bb5fcd74538df3                                                                                                                      0.0s 
 => => exporting manifest list sha256:e1fc0e91df624dbbceb5fec22d2056fac79d8f0a769b9e67466eb1a1efb83706                                                                                                               0.0s 
 => => pushing layers                                                                                                                                                                                                2.7s 
------
 > exporting to image:
------
failed to solve: rpc error: code = Unknown desc = server message: insufficient_scope: authorization failed
```

**Comment by RehanSaeed** [comment]
`docker login` needs to succeed. Also, I've added a `--tag` parameter which is required. See the `build.yml` file for what to pass there.

**Comment by VictorioBerra** [comment]
It does for me:

image

Also, `dotnet cake --push=true --tag="helloworldtory"` results in the same thing. I assume it was not needed because in the code you have `tag = tag ?? dockerfile.GetDirectory().GetDirectoryName().ToLower();`:

```
------
 > exporting to image:
------
failed to solve: rpc error: code = Unknown desc = server message: insufficient_scope: authorization failed
```

**Comment by RehanSaeed** [comment]
Having `--tag="helloworldtory` locally is fine but if you need to push to Docker Hub, you need your username at the start e.g. `--tag=username/helloworldtory`.

**Comment by VictorioBerra** [comment]
That worked!

image

image

**Comment by RehanSaeed** [comment]
Awesome. I'm going to get this merged then.

**Comment by VictorioBerra** [comment]
I think the buildx has a long way to go with the polish though. The error messages are super unhelpful.

**Comment by VictorioBerra** [comment]
I recognize that could just be my lack of thorough understanding of docker and buildx though.

**Comment by RehanSaeed** [comment]
Agree about the error messages.

I've now also added a fallback to use the legacy `docker build` command instead of `docker buildx`. However, I've defaulted to `docker buildx` only because the CI build supports it already and I don't expect devs to run the cake script very often.

**Comment by VictorioBerra** [comment]
> During normal development, VS runs it's own `docker build` commands.

Do you mean via the Docker profile in launchSettings.json?

**Comment by RehanSaeed** [comment]
> 
> 
> > During normal development, VS runs it's own `docker build` commands.
> 
> Do you mean via the Docker profile in launchSettings.json?

Yes indeed!

**Pull request merged**